Sprint 12

urweakness opened the pull request.

urweakness committed:
sprint_12 initial state

urweakness committed:
sprint_12 completed

NaR0RoW commented on ImageFeed/ImageFeed/Models/UnsplashPhotoResult.swift:
```diff
@@ -0,0 +1,50 @@
+    let width: Int
+    let height: Int
+    let description: String?
+    let liked_by_user: Bool
```
В `Swift` принято названия в [`camelCase`](https://www.book2s.com/tutorials/swift-syntax.html#:~:text=they%20enhance%20clarity.-,Naming%20Conventions,-%3A) указывать.
В CodingKeys можно переопределить поле:
`case createdAt = "created_at"`

NaR0RoW commented on ImageFeed/ImageFeed/Presentation/Controllers/TabBarController.swift:
```diff
@@ -1,11 +1,31 @@
+    }
+    
+    // MARK: - Internal View Appearance Configuration
+    override var preferredStatusBarStyle: UIStatusBarStyle {
+        return .darkContent
+    }
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on ImageFeed/ImageFeed/Presentation/ViewControllers/ImageList/ImageListViewController.swift:
```diff
@@ -8,69 +8,177 @@ final class ImageListViewController: UIViewController {
+        alertPresenter = AlertPresenter()
+        
+        UIBlockingActivityIndicator.showActivityIndicator()
+        imagesListService.fetchPhotosNextPage { [weak self] result in
+            guard let self else { return }
+            switch result {
```
Если `self` будет `nil`, то `UIBlockingProgressHUD` не скроется

NaR0RoW commented on ImageFeed/ImageFeed/Presentation/ViewControllers/ImageList/ImageListViewController.swift:
```diff
@@ -8,69 +8,177 @@ final class ImageListViewController: UIViewController {
+            token: token,
+            photoIndex: indexPath.row
+        ) { [weak self] result in
+            guard let self else { return }
+            switch result {
```
Тут тоже если `self` будет `nil`, то `UIBlockingProgressHUD` не скроется

NaR0RoW commented on ImageFeed/ImageFeed/Presentation/ViewControllers/ImageList/ImageListViewController.swift:
```diff
@@ -8,69 +8,177 @@ final class ImageListViewController: UIViewController {
+        forRowAt indexPath: IndexPath
+    ) {
+        if indexPath.row == photos.count - 1 {
+            UIBlockingActivityIndicator.showActivityIndicator()
+            imagesListService.fetchPhotosNextPage { [weak self] result in
+                guard let self else { return }
```
Давай везде исправим момент с сокрытием ProgressHUD в случае self == nil

NaR0RoW commented on ImageFeed/ImageFeed/Presentation/ViewControllers/ImageList/ImagesListCell.swift:
```diff
@@ -28,6 +35,39 @@ final class ImagesListCell: UITableViewCell {
+            case .success:
+                cellImage.contentMode = .scaleAspectFill
+            case .failure:
+                cellImage.image = UIImage(named: "Mark-Stub")
```
Лучше использовать `UIImage(resource: .markStub)`, потому что это безопаснее - такие ресурсы проверяются на этапе компиляции, есть автодополнение, и меньше шансов ошибиться в названии. Строки вроде `"Mark-Stub"` легко написать с опечаткой, и тогда изображение просто не загрузится

NaR0RoW commented on ImageFeed/ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,306 @@
+                }
+            }
+            
+            DispatchQueue.main.async { [weak self] in
+                self?.resumeImagesFetchingTaskIfNeeded()
+            }
```
Безусловной рекомендацией является добавление [weak self] в кложуру, если сомневаешься в поведении.
Но иногда использование [weak self] в кложурах можно опускать, например [внутри GCD](https://temofeev.ru/info/articles/nuzhno-li-pisat-weak-self-v-grand-central-dispatch/), так как это никаким образом не приводит к утечкам.
Так же отличная [статья](https://medium.com/@almalehdev/you-dont-always-need-weak-self-a778bec505ef/) на тему того, когда использование [weak self] можно опустить, для более глубокого изучения, однако ресурс может быть заблокирован.

NaR0RoW commented on ImageFeed/ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,306 @@
+                DispatchQueue.main.async { [weak self] in
+                    guard let self else { return }
+                    switch result {
+                    case .success(let UnsplashPhotoLikeResult):
```
Название констант должно начинаться со строчной буквы

urweakness committed:
fix(sprint_12/ImageListService): ISO8601DateFormatter added once created at singleton instead of UnsplashPhotoResult convert function extension
style(sprint_12): UIImage(named: ) replaced by UIImage(resource: ) for better practice
refactor(sprint_12): Removing excess of using [weak self] capturing in ViewController that instantly execute this closure
